Error with msvc compiler after installing opencl

Hi Michael!

I tried to install opencl but I keep getting the following error when running an example requiring openCL. It does work with the examples that does not require openCL though. It seems like it's related to the msvc compiler so I tried to uninstall and install visual studio 2019 but it keeps giving me the same error. Do you know what might be the problem?

Thanks a lot!

Nicolas

```
[22:01:12] Loading Myokit IDE
[22:01:14] Opened E:\Etudes\Stage ICM\Projet Myokit\examples\decker-2009.mmt
[22:01:28] Running embedded script.
[22:01:34] Done.
[22:01:50] Opened E:\Etudes\Stage ICM\Projet Myokit\OpenCL\stewart-2009-cable-opencl.mmt
[22:01:53] Running embedded script.
[22:01:56] An error has occurred
[22:01:56] Traceback (most recent call last):
  File "c:\users\mbola\miniconda3\lib\distutils\_msvccompiler.py", line 508, in link
    self.spawn([self.linker] + ld_args)
  File "c:\users\mbola\miniconda3\lib\distutils\_msvccompiler.py", line 518, in spawn
    return super().spawn(cmd)
  File "c:\users\mbola\miniconda3\lib\distutils\ccompiler.py", line 910, in spawn
    spawn(cmd, dry_run=self.dry_run)
  File "c:\users\mbola\miniconda3\lib\distutils\spawn.py", line 38, in spawn
    _spawn_nt(cmd, search_path, dry_run=dry_run)
  File "c:\users\mbola\miniconda3\lib\distutils\spawn.py", line 80, in _spawn_nt
    raise DistutilsExecError(
distutils.errors.DistutilsExecError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1146

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\mbola\miniconda3\lib\distutils\core.py", line 148, in setup
    dist.run_commands()
  File "c:\users\mbola\miniconda3\lib\distutils\dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "c:\users\mbola\miniconda3\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "c:\users\mbola\miniconda3\lib\site-packages\setuptools\command\build_ext.py", line 79, in run
    _build_ext.run(self)
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 340, in run
    self.build_extensions()
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 449, in build_extensions
    self._build_extensions_serial()
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 474, in _build_extensions_serial
    self.build_extension(ext)
  File "c:\users\mbola\miniconda3\lib\site-packages\setuptools\command\build_ext.py", line 196, in build_extension
    _build_ext.build_extension(self, ext)
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 550, in build_extension
    self.compiler.link_shared_object(
  File "c:\users\mbola\miniconda3\lib\distutils\ccompiler.py", line 713, in link_shared_object
    self.link(CCompiler.SHARED_OBJECT, objects,
  File "c:\users\mbola\miniconda3\lib\distutils\_msvccompiler.py", line 510, in link
    raise LinkError(msg)
distutils.errors.LinkError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1146

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\__init__.py", line 192, in _compile
    setup(
  File "c:\users\mbola\miniconda3\lib\site-packages\setuptools\__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "c:\users\mbola\miniconda3\lib\distutils\core.py", line 163, in setup
    raise SystemExit("error: " + str(msg))
SystemExit: error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1146

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\gui\ide.py", line 938, in action_run
    myokit.run(
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\_aux.py", line 1134, in run
    r.run()
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\_aux.py", line 1125, in run
    myokit._exec(self.script, environment)
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\_exec_new.py", line 15, in _exec
    exec(script, globals, locals)
  File "<string>", line 14, in <module>
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\openclsim.py", line 320, in __init__
    self._sim = self._compile(
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\__init__.py", line 209, in _compile
    raise myokit.CompilationError('\n'.join(t))
myokit._err.CompilationError: Unable to compile.
Error message:
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1146
Error traceback
Traceback (most recent call last):
  File "c:\users\mbola\miniconda3\lib\distutils\_msvccompiler.py", line 508, in link
    self.spawn([self.linker] + ld_args)
  File "c:\users\mbola\miniconda3\lib\distutils\_msvccompiler.py", line 518, in spawn
    return super().spawn(cmd)
  File "c:\users\mbola\miniconda3\lib\distutils\ccompiler.py", line 910, in spawn
    spawn(cmd, dry_run=self.dry_run)
  File "c:\users\mbola\miniconda3\lib\distutils\spawn.py", line 38, in spawn
    _spawn_nt(cmd, search_path, dry_run=dry_run)
  File "c:\users\mbola\miniconda3\lib\distutils\spawn.py", line 80, in _spawn_nt
    raise DistutilsExecError(
distutils.errors.DistutilsExecError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1146

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\mbola\miniconda3\lib\distutils\core.py", line 148, in setup
    dist.run_commands()
  File "c:\users\mbola\miniconda3\lib\distutils\dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "c:\users\mbola\miniconda3\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "c:\users\mbola\miniconda3\lib\site-packages\setuptools\command\build_ext.py", line 79, in run
    _build_ext.run(self)
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 340, in run
    self.build_extensions()
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 449, in build_extensions
    self._build_extensions_serial()
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 474, in _build_extensions_serial
    self.build_extension(ext)
  File "c:\users\mbola\miniconda3\lib\site-packages\setuptools\command\build_ext.py", line 196, in build_extension
    _build_ext.build_extension(self, ext)
  File "c:\users\mbola\miniconda3\lib\distutils\command\build_ext.py", line 550, in build_extension
    self.compiler.link_shared_object(
  File "c:\users\mbola\miniconda3\lib\distutils\ccompiler.py", line 713, in link_shared_object
    self.link(CCompiler.SHARED_OBJECT, objects,
  File "c:\users\mbola\miniconda3\lib\distutils\_msvccompiler.py", line 510, in link
    raise LinkError(msg)
distutils.errors.LinkError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1146

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\__init__.py", line 192, in _compile
    setup(
  File "c:\users\mbola\miniconda3\lib\site-packages\setuptools\__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "c:\users\mbola\miniconda3\lib\distutils\core.py", line 163, in setup
    raise SystemExit("error: " + str(msg))
SystemExit: error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1146

Compiler output:
    running build_ext
    building 'myokit_sim_opencl_1_2870234442241405037' extension
    creating build
    creating build\temp.win-amd64-3.8
    creating build\temp.win-amd64-3.8\Release
    creating build\temp.win-amd64-3.8\Release\Users
    creating build\temp.win-amd64-3.8\Release\Users\mbola
    creating build\temp.win-amd64-3.8\Release\Users\mbola\AppData
    creating build\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local
    creating build\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local\Temp
    creating build\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD "-IC:\Program Files\Intel\OpenCL SDK\6.3\include" "-IC:\Program Files\AMD APP SDK\2.9\include" "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" "-IC:\Program Files (x86)\IntelSWTools\system_studio_2020\OpenCL\sdk\include" -I -Ic:\users\mbola\miniconda3\lib\site-packages\myokit\_sim -Ic:\users\mbola\miniconda3\lib\site-packages\myokit\_sim -Ic:\users\mbola\miniconda3\include -Ic:\users\mbola\miniconda3\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /TcC:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.c /Fobuild\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.obj
    source.c
    cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
    c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\mcl.h(683): warning C4244: 'fonction'ÿ: conversion de 'cl_ulong' en 'long', perte possible de donn‚es
    c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\mcl.h(690): warning C4244: 'fonction'ÿ: conversion de 'cl_ulong' en 'long', perte possible de donn‚es
    c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\mcl.h(697): warning C4244: 'fonction'ÿ: conversion de 'cl_ulong' en 'long', perte possible de donn‚es
    c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\mcl.h(711): warning C4267: 'fonction'ÿ: conversion de 'size_t' en 'long', perte possible de donn‚es
    c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\mcl.h(727): warning C4267: 'fonction'ÿ: conversion de 'size_t' en 'long', perte possible de donn‚es
    c:\users\mbola\miniconda3\lib\site-packages\myokit\_sim\mcl.h(736): warning C4267: 'fonction'ÿ: conversion de 'size_t' en 'long', perte possible de donn‚es
    C:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.c(397): warning C4244: '='ÿ: conversion de 'Py_ssize_t' en 'int', perte possible de donn‚es
    C:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.c(406): warning C4244: '='ÿ: conversion de 'Py_ssize_t' en 'int', perte possible de donn‚es
    C:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.c(494): warning C4244: '='ÿ: conversion de 'Py_ssize_t' en 'int', perte possible de donn‚es
    C:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.c(751): warning C4244: '='ÿ: conversion de 'Py_ssize_t' en 'int', perte possible de donn‚es
    C:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.c(590): warning C4996: 'clCreateCommandQueue': a ‚t‚ d‚clar‚ d‚conseill‚
    C:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.c(1124): warning C4113: 'PyObject *(__cdecl *)()' est diff‚rent de 'PyCFunction' dans les listes de paramŠtres
    creating C:\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\build\build\lib.win-amd64-3.8
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO "/LIBPATH:C:\Program Files\Intel\OpenCL SDK\6.3\lib\x64" "/LIBPATH:C:\Program Files\AMD APP SDK\2.9\bin\x64" "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64" "/LIBPATH:C:\Program Files (x86)\IntelSWTools\system_studio_2020\OpenCL\sdk\lib\x64" /LIBPATH: /LIBPATH:c:\users\mbola\miniconda3\libs /LIBPATH:c:\users\mbola\miniconda3\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x64" OpenCL.lib /EXPORT:PyInit_myokit_sim_opencl_1_2870234442241405037 build\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.obj /OUT:build\lib.win-amd64-3.8\myokit_sim_opencl_1_2870234442241405037.cp38-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\myokit_sim_opencl_1_2870234442241405037.cp38-win_amd64.lib
    Microsoft (R) Incremental Linker Version 14.28.29336.0
    Copyright (C) Microsoft Corporation.  All rights reserved.
    
    LINK : fatal error LNK1146: aucun argument sp‚cifi‚ avec l'option '/LIBPATH:'
```

The last line of the error message is the most important one here:

```LINK : fatal error LNK1146: aucun argument sp‚cifi‚ avec l'option '/LIBPATH:'```

It means that something is going wrong at the "linking" stage. So visual studio has compiled the simulation code, and now wants to link it to the shared libraries from OpenCL. But, it seems to complain that "no argument" was specified with LIBPATH.

A few lines above that I can see
```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO "/LIBPATH:C:\Program Files\Intel\OpenCL SDK\6.3\lib\x64" "/LIBPATH:C:\Program Files\AMD APP SDK\2.9\bin\x64" "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64" "/LIBPATH:C:\Program Files (x86)\IntelSWTools\system_studio_2020\OpenCL\sdk\lib\x64" /LIBPATH: /LIBPATH:c:\users\mbola\miniconda3\libs /LIBPATH:c:\users\mbola\miniconda3\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x64" OpenCL.lib /EXPORT:PyInit_myokit_sim_opencl_1_2870234442241405037 build\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\source.obj /OUT:build\lib.win-amd64-3.8\myokit_sim_opencl_1_2870234442241405037.cp38-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.8\Release\Users\mbola\AppData\Local\Temp\tmpdu5doah8myokit\myokit_sim_opencl_1_2870234442241405037.cp38-win_amd64.lib
```

which does include this bit:
```
"/LIBPATH:C:\Program Files (x86)\IntelSWTools\system_studio_2020\OpenCL\sdk\lib\x64" 
/LIBPATH: 
/LIBPATH:c:\users\mbola\miniconda3\libs
```

I'm guessing that might be due to the empty spaces in your ini file argument, or perhaps it's because the line ends with `;` ?

If you can remove the trailing `;`s in your `myokit.ini` file and let me know if that fixes it, I can write some code to do this automatically from now on!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@ This page lists the main changes made to Myokit in each release.
 - Fixed
   - [#684](https://github.com/MichaelClerx/myokit/pull/684) Fixed OpenCL loading issue on OS/X (with special thanks to Martin Aguilar and David Augustin).
   - [#686](https://github.com/MichaelClerx/myokit/pull/686) Fixed a (windows only) bug in `myokit.format_path()`.
+  - [#689](https://github.com/MichaelClerx/myokit/pull/689) Path lists read from `myokit.ini` are now filtered for empty entries and closing semicolons.
 
 ## [1.32.0] - 2021-01-19
 - Added

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ This page lists the main changes made to Myokit in each release.
   - [#682](https://github.com/MichaelClerx/myokit/pull/682) OpenCL code now recognises error code -1001.
   - [#683](https://github.com/MichaelClerx/myokit/pull/683) Now testing OpenCL code in CI.
 - Changed
+  - [#689](https://github.com/MichaelClerx/myokit/pull/689) In Python 2, an `ImportError` is now raised if `myokit.ini` contains the sequence " ;" in any of its value (as this cannot be processed by Python 2's `ConfigParser`).
 - Deprecated
 - Removed
   - [#683](https://github.com/MichaelClerx/myokit/pull/683) No longer testing on Python 2.7.6 on linux, or any Python 2.7 on Windows.

--- a/myokit/_config.py
+++ b/myokit/_config.py
@@ -191,7 +191,7 @@ def _load():
 
     # In Python <3.2, strings like "x ; y" are treated as "x" followed by a
     # comment. These shouldn't appear in myokit ini files!
-    if sys.hexversion < 0x03020000:
+    if sys.hexversion < 0x03020000:     # pragma: no cover
         with open(path, 'r') as f:
             lines = f.readlines()
 

--- a/myokit/_config.py
+++ b/myokit/_config.py
@@ -16,6 +16,7 @@ import myokit
 
 # Load libraries
 import os
+import sys
 import logging
 import platform
 
@@ -187,6 +188,25 @@ def _load():
     # No file present? Create one and return
     if not os.path.isfile(path):
         _create(path)
+
+    # In Python <3.2, strings like "x ; y" are treated as "x" followed by a
+    # comment. These shouldn't appear in myokit ini files!
+    if sys.hexversion < 0x03020000:
+        with open(path, 'r') as f:
+            lines = f.readlines()
+
+        import re
+        inline_comment = re.compile('[\w]+[\s]*=[\s]*.+?\s+(;)')
+        for i, line in enumerate(lines):
+            m = inline_comment.match(line)
+            if m is not None:
+                x = m.start(1) - 1
+                raise ImportError(
+                    'Unsupported syntax found in ' + str(path) + ' on line '
+                    + str(1 + i) + ', character ' + str(x) + ', semicolons (;)'
+                    + ' must not be preceded by whitespace: ```'
+                    + line.strip() + '```.')
+        del(lines, inline_comment)
 
     # Create the config parser (no value allows comments)
     config = ConfigParser(allow_no_value=True)

--- a/myokit/_config.py
+++ b/myokit/_config.py
@@ -245,11 +245,9 @@ def _load():
 
     # Sundials libraries, header files, and version
     if config.has_option('sundials', 'lib'):
-        for x in config.get('sundials', 'lib').split(';'):
-            myokit.SUNDIALS_LIB.append(x.strip())
+        myokit.SUNDIALS_LIB.extend(_path_list(config.get('sundials', 'lib')))
     if config.has_option('sundials', 'inc'):
-        for x in config.get('sundials', 'inc').split(';'):
-            myokit.SUNDIALS_INC.append(x.strip())
+        myokit.SUNDIALS_INC.extend(_path_list(config.get('sundials', 'inc')))
     if config.has_option('sundials', 'version'):
         try:
             myokit.SUNDIALS_VERSION = int(config.get('sundials', 'version'))
@@ -278,11 +276,9 @@ def _load():
 
     # OpenCL libraries and header files
     if config.has_option('opencl', 'lib'):
-        for x in config.get('opencl', 'lib').split(';'):
-            myokit.OPENCL_LIB.append(x.strip())
+        myokit.OPENCL_LIB.extend(_path_list(config.get('opencl', 'lib')))
     if config.has_option('opencl', 'inc'):
-        for x in config.get('opencl', 'inc').split(';'):
-            myokit.OPENCL_INC.append(x.strip())
+        myokit.OPENCL_INC.extend(_path_list(config.get('opencl', 'inc')))
 
 
 def _dynamically_add_embedded_sundials_win():   # pragma: no linux cover
@@ -301,6 +297,10 @@ def _dynamically_add_embedded_sundials_win():   # pragma: no linux cover
         myokit.SUNDIALS_LIB.append(os.path.join(sundials_win, 'lib'))
     if len(myokit.SUNDIALS_INC) == 0:
         myokit.SUNDIALS_INC.append(os.path.join(sundials_win, 'include'))
+
+
+def _path_list(text):
+    return [x for x in [x.strip() for x in text.split(';')] if x != '']
 
 
 # Load settings

--- a/myokit/tests/test_config.py
+++ b/myokit/tests/test_config.py
@@ -41,21 +41,21 @@ inc = three;eight
 # Config with empty paths and spaces
 config_empties_1 = """
 [sundials]
-lib =
+lib = five
 inc = three;four;
 [opencl]
 lib = one;;   two point five;three;
-inc = five
+inc =
 """
 
 # Config with empty paths and " ;", which in Python 2 is ignored
 config_empties_2 = """
 [sundials]
-lib =
+lib = five ; six
 inc = three;four;
 [opencl]
 lib = one ;;   two point five ;three;
-inc = five ; six
+inc =
 """
 
 # Qt options
@@ -174,12 +174,12 @@ class TestConfig(unittest.TestCase):
                 with open(d.path('myokit.ini'), 'w') as f:
                     f.write(config_empties_1)
                 config._load()
-                self.assertEqual(myokit.SUNDIALS_LIB, [])
+                self.assertEqual(myokit.SUNDIALS_LIB, ['five'])
                 self.assertEqual(myokit.SUNDIALS_INC, ['three', 'four'])
                 self.assertEqual(
                     myokit.OPENCL_LIB,
                     ['one', 'two point five', 'three'])
-                self.assertEqual(myokit.OPENCL_INC, ['five'])
+                self.assertEqual(myokit.OPENCL_INC, [])
 
                 # Even if the list contains " ;", which Python 2's config
                 # parser treats as a comment
@@ -193,12 +193,12 @@ class TestConfig(unittest.TestCase):
                     self.assertRaises(ImportError, config._load)
                 else:
                     config._load()
-                    self.assertEqual(myokit.SUNDIALS_LIB, [])
+                    self.assertEqual(myokit.SUNDIALS_LIB, ['five', 'six'])
                     self.assertEqual(myokit.SUNDIALS_INC, ['three', 'four'])
                     self.assertEqual(
                         myokit.OPENCL_LIB,
                         ['one', 'two point five', 'three'])
-                    self.assertEqual(myokit.OPENCL_INC, ['five', 'six'])
+                    self.assertEqual(myokit.OPENCL_INC, [])
 
                 # Qt gui options
                 with open(d.path('myokit.ini'), 'w') as f:

--- a/myokit/tests/test_config.py
+++ b/myokit/tests/test_config.py
@@ -39,13 +39,23 @@ inc = three;eight
 """
 
 # Config with empty paths and spaces
-config_empties = """
+config_empties_1 = """
 [sundials]
 lib =
-inc = three;
+inc = three;four;
+[opencl]
+lib = one;;   two point five;three;
+inc = five
+"""
+
+# Config with empty paths and " ;", which in Python 2 is ignored
+config_empties_2 = """
+[sundials]
+lib =
+inc = three;four;
 [opencl]
 lib = one ;;   two point five ;three;
-inc = five
+inc = five ; six
 """
 
 # Qt options
@@ -155,28 +165,6 @@ class TestConfig(unittest.TestCase):
                 self.assertEqual(myokit.OPENCL_LIB, ['five', 'six'])
                 self.assertEqual(myokit.OPENCL_INC, ['three', 'eight'])
 
-                # Full values, PySide gui
-                myokit.SUNDIALS_LIB = []
-                myokit.SUNDIALS_INC = []
-                myokit.OPENCL_LIB = []
-                myokit.OPENCL_INC = []
-                myokit.FORCE_PYSIDE = myokit.FORCE_PYSIDE2 = False
-                myokit.FORCE_PYQT4 = myokit.FORCE_PYQT5 = False
-                with open(d.path('myokit.ini'), 'w') as f:
-                    f.write(config2)
-                config._load()
-                self.assertEqual(myokit.DATE_FORMAT, 'TEST_DATE_FORMAT')
-                self.assertEqual(myokit.TIME_FORMAT, 'TEST_TIME_FORMAT')
-                self.assertTrue(myokit.DEBUG_LINE_NUMBERS)
-                self.assertFalse(myokit.FORCE_PYSIDE)
-                self.assertFalse(myokit.FORCE_PYSIDE2)
-                self.assertFalse(myokit.FORCE_PYQT4)
-                self.assertFalse(myokit.FORCE_PYQT5)
-                self.assertEqual(myokit.SUNDIALS_LIB, ['one', 'two'])
-                self.assertEqual(myokit.SUNDIALS_INC, ['three', 'four'])
-                self.assertEqual(myokit.OPENCL_LIB, ['five', 'six'])
-                self.assertEqual(myokit.OPENCL_INC, ['three', 'eight'])
-
                 # Lists of paths should be filtered for empty values and
                 # trimmed
                 myokit.SUNDIALS_LIB = []
@@ -184,14 +172,33 @@ class TestConfig(unittest.TestCase):
                 myokit.OPENCL_LIB = []
                 myokit.OPENCL_INC = []
                 with open(d.path('myokit.ini'), 'w') as f:
-                    f.write(config_empties)
+                    f.write(config_empties_1)
                 config._load()
                 self.assertEqual(myokit.SUNDIALS_LIB, [])
-                self.assertEqual(myokit.SUNDIALS_INC, ['three'])
+                self.assertEqual(myokit.SUNDIALS_INC, ['three', 'four'])
                 self.assertEqual(
                     myokit.OPENCL_LIB,
                     ['one', 'two point five', 'three'])
                 self.assertEqual(myokit.OPENCL_INC, ['five'])
+
+                # Even if the list contains " ;", which Python 2's config
+                # parser treats as a comment
+                myokit.SUNDIALS_LIB = []
+                myokit.SUNDIALS_INC = []
+                myokit.OPENCL_LIB = []
+                myokit.OPENCL_INC = []
+                with open(d.path('myokit.ini'), 'w') as f:
+                    f.write(config_empties_2)
+                if sys.hexversion < 0x03020000:
+                    self.assertRaises(ImportError, config._load)
+                else:
+                    config._load()
+                    self.assertEqual(myokit.SUNDIALS_LIB, [])
+                    self.assertEqual(myokit.SUNDIALS_INC, ['three', 'four'])
+                    self.assertEqual(
+                        myokit.OPENCL_LIB,
+                        ['one', 'two point five', 'three'])
+                    self.assertEqual(myokit.OPENCL_INC, ['five', 'six'])
 
                 # Qt gui options
                 with open(d.path('myokit.ini'), 'w') as f:

--- a/myokit/tests/test_config.py
+++ b/myokit/tests/test_config.py
@@ -38,6 +38,16 @@ lib = five;six
 inc = three;eight
 """
 
+# Config with empty paths and spaces
+config_empties = """
+[sundials]
+lib =
+inc = three;
+[opencl]
+lib = one ;;   two point five ;three;
+inc = five
+"""
+
 # Qt options
 config_pyside = """
 [gui]
@@ -144,6 +154,44 @@ class TestConfig(unittest.TestCase):
                 self.assertEqual(myokit.SUNDIALS_INC, ['three', 'four'])
                 self.assertEqual(myokit.OPENCL_LIB, ['five', 'six'])
                 self.assertEqual(myokit.OPENCL_INC, ['three', 'eight'])
+
+                # Full values, PySide gui
+                myokit.SUNDIALS_LIB = []
+                myokit.SUNDIALS_INC = []
+                myokit.OPENCL_LIB = []
+                myokit.OPENCL_INC = []
+                myokit.FORCE_PYSIDE = myokit.FORCE_PYSIDE2 = False
+                myokit.FORCE_PYQT4 = myokit.FORCE_PYQT5 = False
+                with open(d.path('myokit.ini'), 'w') as f:
+                    f.write(config2)
+                config._load()
+                self.assertEqual(myokit.DATE_FORMAT, 'TEST_DATE_FORMAT')
+                self.assertEqual(myokit.TIME_FORMAT, 'TEST_TIME_FORMAT')
+                self.assertTrue(myokit.DEBUG_LINE_NUMBERS)
+                self.assertFalse(myokit.FORCE_PYSIDE)
+                self.assertFalse(myokit.FORCE_PYSIDE2)
+                self.assertFalse(myokit.FORCE_PYQT4)
+                self.assertFalse(myokit.FORCE_PYQT5)
+                self.assertEqual(myokit.SUNDIALS_LIB, ['one', 'two'])
+                self.assertEqual(myokit.SUNDIALS_INC, ['three', 'four'])
+                self.assertEqual(myokit.OPENCL_LIB, ['five', 'six'])
+                self.assertEqual(myokit.OPENCL_INC, ['three', 'eight'])
+
+                # Lists of paths should be filtered for empty values and
+                # trimmed
+                myokit.SUNDIALS_LIB = []
+                myokit.SUNDIALS_INC = []
+                myokit.OPENCL_LIB = []
+                myokit.OPENCL_INC = []
+                with open(d.path('myokit.ini'), 'w') as f:
+                    f.write(config_empties)
+                config._load()
+                self.assertEqual(myokit.SUNDIALS_LIB, [])
+                self.assertEqual(myokit.SUNDIALS_INC, ['three'])
+                self.assertEqual(
+                    myokit.OPENCL_LIB,
+                    ['one', 'two point five', 'three'])
+                self.assertEqual(myokit.OPENCL_INC, ['five'])
 
                 # Qt gui options
                 with open(d.path('myokit.ini'), 'w') as f:

